x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecyclePolicyMetadata.java

```diff
@@ -18,20 +18,70 @@
 
 import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 
 public class LifecyclePolicyMetadata extends AbstractDiffable<LifecyclePolicyMetadata>
         implements ToXContentObject, Diffable<LifecyclePolicyMetadata> {
 
+    public static final ParseField TYPE_FIELD = new ParseField("type");
     public static final ParseField POLICY = new ParseField("policy");
     public static final ParseField HEADERS = new ParseField("headers");
     @SuppressWarnings("unchecked")
     public static final ConstructingObjectParser<LifecyclePolicyMetadata, String> PARSER = new ConstructingObjectParser<>("policy_metadata",
-            a -> new LifecyclePolicyMetadata((LifecyclePolicy) a[0], (Map<String, String>) a[1]));
+            a -> {
+                LifecycleType type = (LifecycleType) a[0];
+                LifecyclePolicy dummyPolicy = (LifecyclePolicy) a[1];
+                LifecyclePolicy policy = new LifecyclePolicy(type, dummyPolicy.getName(), dummyPolicy.getPhases());
+                return new LifecyclePolicyMetadata(policy, (Map<String, String>) a[2]);
+            });
     static {
-        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> LifecyclePolicy.parse(p, c), POLICY);
-        PARSER.declareField(ConstructingObjectParser.constructorArg(), p -> p.mapStrings(), HEADERS, ValueType.OBJECT);
+        // this class is here for parsing purposes
+        LifecycleType dummyType = new LifecycleType() {
+            @Override
+            public List<Phase> getOrderedPhases(Map<String, Phase> phases) {
+                return null;
+            }
+
+            @Override
+            public String getNextPhaseName(String currentPhaseName, Map<String, Phase> phases) {
+                return null;
+            }
+
+            @Override
+            public String getPreviousPhaseName(String currentPhaseName, Map<String, Phase> phases) {
+                return null;
+            }
+
+            @Override
+            public List<LifecycleAction> getOrderedActions(Phase phase) {
+                return null;
+            }
+
+            @Override
+            public String getNextActionName(String currentActionName, Phase phase) {
+                return null;
+            }
+
+            @Override
+            public void validate(Collection<Phase> phases) {
+            }
+
+            @Override
+            public String getWriteableName() {
+                return null;
+            }
+
+            @Override
+            public void writeTo(StreamOutput out) {
+            }
+        };
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> p.namedObject(LifecycleType.class, p.text(), null),
+            TYPE_FIELD, ValueType.STRING);
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> LifecyclePolicy.parse(p, c, dummyType), POLICY);
```

Comment:
I hate almost everything about this... but I found this to be better than making the parsing way more complicated. 

The problem here is that `type` was lifted into the MetaData level, but it is not necessarily parsed before LifecyclePolicy can go ahead and parse itself with the `type` in its context, like it requires.

One option was to allow type to be `null` here, but I didn't like what that meant with validation, so I made a dummy type, just for creation.

Reply:
Why does the lifecycle type need to be lifted up into the metadata?